flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/AntiJoin.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.maintenance.operator;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.functions.OpenContext;
+import org.apache.flink.api.common.state.MapState;
+import org.apache.flink.api.common.state.MapStateDescriptor;
+import org.apache.flink.api.common.state.ValueState;
+import org.apache.flink.api.common.state.ValueStateDescriptor;
+import org.apache.flink.api.common.typeinfo.Types;
+import org.apache.flink.streaming.api.functions.co.KeyedCoProcessFunction;
+import org.apache.flink.util.Collector;
+import org.apache.iceberg.actions.DeleteOrphanFiles;
+import org.apache.iceberg.actions.FileURI;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Internal
+public class AntiJoin extends KeyedCoProcessFunction<String, String, String, String> {
+  private static final Logger LOG = LoggerFactory.getLogger(AntiJoin.class);
+
+  private transient MapState<String, Boolean> foundInTable;
```

Comment:
@mxm: We use `MapState` here to dedupe the strings found in the table. Is this the best way to do it in Flink?

@Guosmilesmile: If we decide to use this, we still should leave a comment about the reason we are using MapState here.

Reply:
The MapState principally is fine. Alternatively, we could have used an time window here, where Flink would also internally use state. But this is a custom join operator, so directly using state is fine - as long as the  state eviction policy works. We register a timer for every element's timestamp. The timer ensures the state will be cleared on watermark arrival.